user_linux_cgo.go

```diff
@@ -0,0 +1,216 @@
+			}
+			// Start over with that resulting path as our input.
+			return newfilename, nil, nil
+		}
+
+		// If it's not a directory or a link, and we still have components, it's an invalid path.
+		if len(remaining) > 0 {
+			return "", nil, fmt.Errorf("%q is not a directory", filepath.Join(rootdir, filepath.Join(append(parents, comp)...)))
+		}
+
+		// It's an ordinary file.  Recompute the filename for display and open the file for real, and we're done.
+		fd, err = unix.Openat(fd, comp, unix.O_NOFOLLOW, 0)
+		if err != nil {
+			return "", nil, err
+		}
+		filename = filepath.Join(append(parents, comp)...)
+	}
+
+	// Actually open the file for reading.
+	mode := C.CString("r")
+	defer C.free(unsafe.Pointer(mode))
+	f, err := C.fdopen(C.int(fd), mode)
+	if f == nil || err != nil {
+		return "", nil, fmt.Errorf("error opening %q/%q: %v", rootdir, filename, err)
+	}
+	return filename, f, nil
+}
+
+func fopenContainerFile(rootdir, filename string) (C.pFILE, error) {
+	symlinksEncountered := 0
+	newfilename, file, err := fopenContainerFileWork(rootdir, filename, symlinksEncountered)
+	for file == nil && err == nil {
+		symlinksEncountered++
+		newfilename, file, err = fopenContainerFileWork(rootdir, newfilename, symlinksEncountered)
+	}
+	return file, err
+}
+
+var (
+	lookupUser, lookupGroup sync.Mutex
+)
+
+func lookupUserInContainer(rootdir, username string) (uint64, uint64, error) {
+	name := C.CString(username)
+	defer C.free(unsafe.Pointer(name))
+
+	f, err := fopenContainerFile(rootdir, "/etc/passwd")
+	if err != nil {
+		return 0, 0, err
+	}
+	defer C.fclose(f)
+
+	lookupUser.Lock()
+	defer lookupUser.Unlock()
+
+	pwd := C.fgetpwent(f)
+	for pwd != nil {
+		if C.strcmp(pwd.pw_name, name) != 0 {
+			pwd = C.fgetpwent(f)
+			continue
+		}
+		return uint64(pwd.pw_uid), uint64(pwd.pw_gid), nil
+	}
+
+	return 0, 0, user.UnknownUserError(fmt.Sprintf("error looking up user %q", username))
+}
+
+func lookupGroupForUIDInContainer(rootdir string, userid uint64) (string, uint64, error) {
+	f, err := fopenContainerFile(rootdir, "/etc/passwd")
+	if err != nil {
+		return "", 0, err
+	}
+	defer C.fclose(f)
+
+	lookupUser.Lock()
+	defer lookupUser.Unlock()
+
+	pwd := C.fgetpwent(f)
+	for pwd != nil {
+		if uint64(pwd.pw_uid) != userid {
+			pwd = C.fgetpwent(f)
+			continue
+		}
+		return C.GoString(pwd.pw_name), uint64(pwd.pw_gid), nil
+	}
+
+	return "", 0, user.UnknownUserError(fmt.Sprintf("error looking up user with UID %d", userid))
+}
+
+func lookupGroupInContainer(rootdir, groupname string) (uint64, error) {
+	name := C.CString(groupname)
+	defer C.free(unsafe.Pointer(name))
+
+	f, err := fopenContainerFile(rootdir, "/etc/group")
+	if err != nil {
+		return 0, err
+	}
+	defer C.fclose(f)
+
+	lookupGroup.Lock()
+	defer lookupGroup.Unlock()
+
+	grp := C.fgetgrent(f)
+	for grp != nil {
+		if C.strcmp(grp.gr_name, name) != 0 {
+			grp = C.fgetgrent(f)
```

Comment:
Infinite loop?

Reply:
It shouldn't?  We either (read another entry, and either iterate or fall out of the loop) or (return).

Reply:
Yup, me being dumb.